### ISSUE TITLE
Repositions cursor at bottom of file on untabify

### ISSUE THREAD

**Issue by robhicks** [lee-dohm/tabs-to-spaces#41]
Every time I untabify, the cursor is repositioned at the bottom of the file. Is there a way configure the package to leave the cursor at its current position?

**Comment by lee-dohm** [comment]
I'll take a look at this later today.

**Comment by lee-dohm** [comment]
Sorry it has taken me so long to get back to this. I can't reproduce this using Atom v1.4.1 on Mac OS X 10.11.3 with the latest version of tabs-to-spaces. Could you give the following information?

* Which version of Atom are you using?
* What OS and version are you running?
* Is it certain file types that this happens in?
* What other packages do you have installed?

**Comment by robhicks** [comment]
@lee-dohm I recently updated to Atom v1.4.1 on OSX 10.11.3. Prior to the update, it was happening on all files. Unfortunately, before updating to Atom v1.4.1, I also tweaked some of my settings which resulted in Atom not inserting tabs instead of spaces so I can't reproduce the problem.

**Comment by lee-dohm** [comment]
I'm glad it is working for you now. I'll see if I can find the time to poke around and reproduce the problem. Thanks for reporting it!

**Comment by robhicks** [comment]
@lee-dohm I figured out how to reproduce this. 

If you use atom beautify on a fle, it changes spaces to tabs but leaves the cursor where it was before running it. Then when untabify is run on the file, it changes the tabs to spaces and the cursor is reposition ed at the end of the file.

**Comment by lee-dohm** [comment]
I think you're going to need to be a little more specific. I installed atom-beautify and ran it on a random JS file and couldn't replicate what you're describing using Atom v1.6.0-dev-52c9067 on Mac OS X 10.11.3. Could you give the following information?

* What language is the file you reproduced the issue in?
* Do you have any specific beautify configuration?
* Why do you have beautify change spaces to tabs only to change them back again? :wink:

**Comment by robhicks** [comment]
It happens for all the file types I have tried. I see it mostly with JavaScript files because I spend more time in them. 

I have had the default configuration of Atom Beautify. I have used it mostly to reformat code that get's pasted in incorrectly. 

Your question about why I have Atom Beautify change the indents to tabs and then untabify them was a good one. I have now modified by Atom Beautify settings to use tabs. 

I am using Atom 1.4.1 on Mac OSX 10.11.13. I have the following packages installed:

├── Stylus@1.2.1
├── Sublime-Style-Column-Selection@1.5.1
├── angularjs@0.3.4
├── atom-beautify@0.28.21
├── atom-easy-jsdoc@4.2.1
├── atom-jade@0.3.0
├── atom-tsd@0.9.0
├── atom-typescript@8.2.0
├── autocomplete-paths@1.0.2
├── clipboard-plus@0.5.1
├── color-picker@2.1.1
├── compare-files@0.6.2
├── editorconfig@1.2.4
├── file-icons@1.6.14
├── firepad@0.5.0
├── git-history@3.2.0
├── git-plus@5.12.1
├── highlight-line@0.11.0
├── hyperclick@0.0.35
├── imdone-atom@1.3.25
├── jade-beautify@0.1.0
├── js-hyperclick@1.4.1
├── language-ejs@0.2.0
├── linter@1.11.3
├── linter-csslint@1.2.0
├── linter-docker@0.1.2
├── linter-eslint@6.0.0
├── linter-jade@0.3.2
├── linter-javac@1.5.0
├── linter-jsonlint@1.1.4
├── linter-stylint@2.0.1
├── linter-tslint@0.6.0
├── live-archive@0.1.15
├── markdown-folder@0.5.0
├── markdown-pdf@1.3.9
├── markdown-scroll-sync@2.0.3
├── markdown-themeable-pdf@0.10.2
├── markdown-toc@0.4.1
├── markdown-writer@2.2.0
├── merge-conflicts@1.3.7
├── minimap@4.19.0
├── monokai@0.18.0
├── monokai-seti@0.7.0
├── node-debugger@1.3.2
├── nuclide-click-to-symbol@0.0.22
├── polymer-snippets@3.1.0
├── project-manager@2.7.5
├── rest-client@0.5.0
├── selection-highlight@0.1.4
├── snippets-jade@0.1.1
├── sort-lines@0.14.0
├── split-diff@0.5.4
├── stylus-autocompile@0.3.1
├── svg-preview@0.7.3
├── tabs-to-spaces@1.0.1
└── terminal-plus@0.14.5

**Comment by lee-dohm** [comment]
I was able to reproduce this with both atom-beautify set to use tabs with JS and simply using tabify and then untabify on Atom v1.4.1 on Mac OS X 10.11.3. I'll work on it this week.

**Comment by lee-dohm** [comment]
Sorry this has gone so long without updates. I tracked down the regression to 4de910566e4d64c58348dd11edf09b2f6522166d. The performance optimization I added to resolve #16 prevents accurately retaining cursor positions. Because it replaces all text in the file whenever anything is changed, all cursors are pushed to the end.

In a tradeoff between correctness and performance, my personal opinion is to err on the side of correctness. I'll be reverting the performance optimization ... but hoping that I can retain it as an option in a future version in case people don't care where their cursors end up so long as it is fast :grinning:
